### PR TITLE
Support switch control in forms (BS4 only)

### PR DESCRIPTION
**Pull request by lindyhopchris**
Adds a `switch` control type, for the Bootstrap 4 switch custom control type:
https://getbootstrap.com/docs/4.6/components/forms/#switches

See #912

### CODE REVIEW

**Comment by lindyhopchris** [comment]
FYI I couldn't preview the changes to the docs site as it's not working when I run `ember s` locally in the docs folder - getting an error.

File: tests/integration/components/bs-form/element/control/switch-test.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import { testBS4 } from '../../../../../helpers/bootstrap';
+import hbs from 'htmlbars-inline-precompile';
+import setupNoDeprecations from '../../../../../helpers/setup-no-deprecations';
+
+module('Integration | Component | bs form/element/control/switch', function (hooks) {
+  setupRenderingTest(hooks);
+  setupNoDeprecations(hooks);
+
+  testBS4('it renders', async function (assert) {
+    await render(hbs`<BsForm::Element::Control::Switch />`);
```
Comment (jelhan) [inline tests/integration/components/bs-form/element/control/switch-test.js:13]: As above I would prefer to only use public API in tests:

```suggestion
    await render(hbs`
      <BsForm as |form|>
        <form.element @controlType="switch" />
      </BsForm>
    `);
```

File: tests/integration/components/bs-form/element/label-test.js [hunk 2]
```diff
@@ -42,6 +42,14 @@ module('Integration | Component | bs form/element/label', function (hooks) {
     assert.dom('label').hasClass('form-check-label', 'component has form-check-label class');
   });
 
+  testBS4('component has custom-control-label class when using control type switch', async function (assert) {
+    await render(hbs`<BsForm::Element::Label @controlType="switch" />`);
```
Comment (jelhan) [inline tests/integration/components/bs-form/element/label-test.js:46]: As above I would prefer to only use public API in tests:

```suggestion
    await render(hbs`
      <BsForm as |form|>
        <form.element @controlType="switch" />
      </BsForm>
    `);
```

File: tests/integration/components/bs-form/element-test.js [hunk 3]
```diff
@@ -148,6 +148,22 @@ module('Integration | Component | bs-form/element', function (hooks) {
     });
   });
 
+  testBS4('controlType "switch" is supported', async function (assert) {
+    await render(hbs`<BsForm::Element @controlType="switch" />`);
```
Comment (jelhan) [inline tests/integration/components/bs-form/element-test.js:152]: I know that this style is used at many existing tests. But would prefer to invoke it using public API only. Do you mind changing?

```suggestion
    await render(hbs`
      <BsForm as |form|>
        <form.element @controlType="switch" />
      </BsForm>
    `);
```

**Review by jelhan — COMMENTED** [review]
Thanks a lot for working on this. Haven't had the time yet to do a detailed review. But what I have seen so far looks very good to me. Will try to finish review later today.

> FYI I couldn't preview the changes to the docs site as it's not working when I run `ember s` locally in the docs folder - getting an error.

The docs site is broken since quiet some time. :disappointed: See #1387 for details.

File: addon/components/bs-form/element/control/switch.hbs [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+  type="checkbox"
+  id={{@id}}
+  disabled={{this.disabled}}
+  readonly={{this.readonly}}
+  aria-describedby={{this.ariaDescribedBy}}
+  checked={{this.value}}
+  class="custom-control-input {{this.formValidationClass}}"
+  ...attributes
+  {{on "click" this.handleClick}}
+>
```
Comment (jelhan) [inline addon/components/bs-form/element/control/switch.hbs:11]: Wondering if it makes sense to refactor to:

```hbs
<BsForm::Element::Control::Checkbox
  @value={{@value}}
  @id={{@id}}
  @type="checkbox"
  @label={{@label}}
  @disabled={{@disabled}}
  @readonly={{@readonly}}
  @required={{@required}}
  @options={{@options}}
  @optionLabelPath={{@optionLabelPath}}
  @ariaDescribedBy={{@ariaDescribedBy}}
  @onChange={{@onChange}}
  @validationType={{@validationType}}
  @size={{@size}}
  class="custom-control-input"
  ...attributes
/>
```

We would avoid some code duplication. But needing to explicitly pass down all arguments is a big trade-off. We could only passthrough the arguments actually used by `<BsForm::Element::Control::Checkbox />`. But that would be also error prune...

So not a request to change. More a question on long-term architectural patterns for @simonihmig. :smile_cat:

**Review by jelhan — APPROVED** [review]
Looks good to me. Thanks a lot for working on this one!

File: addon/components/bs-form/element/control/switch.hbs (hunk 4 above)
Comment (lindyhopchris) [reply addon/components/bs-form/element/control/switch.hbs:11]: The problem is that the checkbox control adds a `form-check-input` class, which we do not want on the switch checkbox. Which is why I went with duplicating the control.

Comment (jelhan) [reply addon/components/bs-form/element/control/switch.hbs:11]: Missed that one. Thanks for explaining.

**Comment by lindyhopchris** [comment]
@jelhan thanks for reviewing. I'll make those changes to the test when I get a chance (hopefully later today)

File: addon/components/bs-form/element/control/switch.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import deprecateSubclassing from 'ember-bootstrap/utils/deprecate-subclassing';
+
+/**
+
+ @class FormElementControlSwitch
+ @namespace Components
+ @extends Components.FormElementControl
+ @private
+ */
+@tagName('')
```
Comment (simonihmig) [inline addon/components/bs-form/element/control/switch.js:13]: Hm, reminds me that we still have quite some work left over to move things to Glimmer components. It's ok to do this here as anything else would be out of scope for this PR, but still a bit unfortunate that we basically add additional tech debt here. Again, not your fault!

File: addon/components/bs-form/element/layout/horizontal/checkbox.hbs [hunk 6]
```diff
@@ -8,11 +8,11 @@
     {{component @errorsComponent}}
     {{component @helpTextComponent}}
   {{else}}
-    <div class="form-check">
+    <div class="{{if this.isSwitch "custom-control custom-switch" "form-check"}}">
```
Comment (simonihmig) [inline addon/components/bs-form/element/layout/horizontal/checkbox.hbs:11]: I would suggest to express the same without introducing a CP in the backing class, by using the `bs-eq` helper here. Would ease the eventual transition to a template-only component later. Same for the other components.

In general we started to move more of the very basic derived state like this into the template, to reduce the need to add trivial CPs/getters to the class, again a.o. with the goal to use template-only components wherever possible.

File: tests/integration/components/bs-form/element-test.js (hunk 3 above)
Comment (simonihmig) [reply tests/integration/components/bs-form/element-test.js:152]: I agree. Again not your fault, as examples like this do exist in the existing code base. Which I would consider bad practice nowadays, and should get refactored over time (e.g. in #1455)

**Review by simonihmig — COMMENTED** [review]
Thanks a lot for working on this!

**Comment by lindyhopchris** [comment]
@simonihmig thanks for your comments! All make sense, so I'll be able to push an update. Today is looking busy so hopefully I'll have time tomorrow.

**Comment by lindyhopchris** [comment]
@jelhan @simonihmig

I've pushed the changes based on the review comments, so this should now be ready to merge. Do you need me to fix the merge conflicts? I tried back merging `master` into my `switch` branch but then all the tests started to fail.

**Comment by simonihmig** [comment]
> Do you need me to fix the merge conflicts? 

Yes, unfortunately I cannot merge it with this conflict. I did a few things here yesterday, so sorry for causing this additional work! 😔

> I tried back merging master into my switch branch but then all the tests started to fail.

Hm, what exactly was the error? Did you update your dependencies and restarted `ember serve`?

**Comment by lindyhopchris** [comment]
@simonihmig have pushed the merge fix. Turned out the failing tests were due to a bad merge on my part, re-did the merge and worked out how to fix it, so all should be good now.

**Review by simonihmig — APPROVED** [review]
Thanks a lot again!

**Pull request merged**